src/objective/rank_obj.cu

```diff
@@ -242,98 +242,98 @@ struct MAPLambdaWeightComputer {
+      dlabels_(labels + begin, labels + end),
+      orig_dpreds_(preds),
+      orig_dlabels_(labels),
+      dpos_(end - begin),
+      dlabels_count_(end - begin + 1, 1),
+      begin_group_idx_(begin),
+      device_id_(dev_id) {
+    dh::safe_cuda(cudaStreamCreate(&stream_));
+    thrust::sequence(thrust::cuda::par.on(stream_), dpos_.begin(), dpos_.end());
+    dlabels_count_[0] = 0;
   }
 
   ~SortedLabelList() {
-    dh::safe_cuda(cudaSetDevice(device_id));
-    dh::safe_cuda(cudaStreamDestroy(stream));
+    dh::safe_cuda(cudaSetDevice(device_id_));
+    dh::safe_cuda(cudaStreamDestroy(stream_));
   }
 
   // Sort by predictions first and then sort the labels by predictions next
   void Sort() {
-    dh::device_vector<bst_float> cdpreds(dpreds);
+    dh::device_vector<bst_float> cdpreds(dpreds_);
     // Sort the predictions first and rearrange its positional indices
-    thrust::sort_by_key(thrust::cuda::par.on(stream),
-                        cdpreds.begin(), cdpreds.end(), dpos.begin(),
+    thrust::sort_by_key(thrust::cuda::par.on(stream_),
+                        cdpreds.begin(), cdpreds.end(), dpos_.begin(),
                         thrust::greater<bst_float>());
     // Gather the labels based on the sorted indices
-    thrust::gather(thrust::cuda::par.on(stream),
-                   dpos.begin(), dpos.end(), orig_dlabels + begin_group_idx, dlabels.begin());
+    thrust::gather(thrust::cuda::par.on(stream_),
+                   dpos_.begin(), dpos_.end(), orig_dlabels_ + begin_group_idx_, dlabels_.begin());
 
     // Sort the labels next and get the final order
-    thrust::sort_by_key(thrust::cuda::par.on(stream),
-                        dlabels.begin(), dlabels.end(), dpos.begin(),
+    thrust::sort_by_key(thrust::cuda::par.on(stream_),
+                        dlabels_.begin(), dlabels_.end(), dpos_.begin(),
                         thrust::greater<bst_float>());
     // Use the order to then sort the original predictions
-    thrust::gather(thrust::cuda::par.on(stream),
-                   dpos.begin(), dpos.end(), orig_dpreds + begin_group_idx, dpreds.begin());
+    thrust::gather(thrust::cuda::par.on(stream_),
+                   dpos_.begin(), dpos_.end(), orig_dpreds_ + begin_group_idx_, dpreds_.begin());
   }
 
   // For all the unique labels, create a number of such labels for those unique label
   // values. Returns the number of such unique labels
   int CreateUniqueLabelCount() {
-    dh::device_vector<bst_float> dunique_labels(dlabels.size());
+    dh::device_vector<bst_float> dunique_labels(dlabels_.size());
     // Find all unique values first and the number of such labels in that group
     auto itr_pair =
-      thrust::reduce_by_key(thrust::cuda::par.on(stream),
-                            dlabels.begin(), dlabels.end(), dlabels_count.begin() + 1,
-                            dunique_labels.begin(), dlabels_count.begin() + 1);
+      thrust::reduce_by_key(thrust::cuda::par.on(stream_),
```

Comment:
This function has very poor performance typically because it has to dynamically work out how many different keys there are. If you know the number of keys to expect cub segmented reduce can give you much better performance.

Reply:
n/a anymore